exonum-java-binding/core/src/main/java/com/exonum/binding/core/storage/database/Prefixed.java

```diff
@@ -0,0 +1,106 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.core.storage.database;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.exonum.binding.core.proxy.Cleaner;
+import com.exonum.binding.core.proxy.NativeHandle;
+import com.exonum.binding.core.proxy.ProxyDestructor;
+import com.exonum.binding.core.storage.indices.IndexAddress;
+import com.exonum.binding.core.util.LibraryLoader;
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * A prefixed database access. It uses a base Access, and adds an address resolution.
+ *
+ * <p>The Prefixed Access resolves the index addresses by prepending a namespace, followed by
+ * a dot ('.'), to the {@linkplain IndexAddress#getName() name part} of the address.
+ *
+ * <p>This class is a native proxy of the {@code Prefixed} Rust Access.
+ */
+public final class Prefixed extends AbstractAccess {
+
+  static {
+    LibraryLoader.load();
+  }
+
+  private final Cleaner cleaner;
+
+  Prefixed(NativeHandle prefixedNativeHandle, boolean canModify, Cleaner cleaner,
+      OpenIndexRegistry registry) {
+    super(prefixedNativeHandle, canModify, registry);
+    this.cleaner = cleaner;
+  }
+
+  /**
+   * Creates a new Prefixed access given the base database access and the namespace.
+   *
+   * <p>The new Prefixed access will inherit from the base access its "canModify" property;
+   * and use its cleaner and registry of open indexes. When that cleaner gets closed,
+   * this access will also be closed.
+   *
+   * @param namespace the namespace to use
+   * @param baseAccess the base database access
+   */
+  @VisibleForTesting static Prefixed fromAccess(String namespace, AbstractAccess baseAccess) {
+    // todo: If it is not used beyond tests, consider removing the _registry sharing_,
+    //   since that's the only method that uses (and needs) shared index registry.
+    Cleaner cleaner = baseAccess.getCleaner();
+    OpenIndexRegistry registry = baseAccess.getOpenIndexes();
+    long handle = nativeCreate(namespace, baseAccess.getAccessNativeHandle());
+    return fromHandleInternal(handle, baseAccess.canModify(), cleaner, registry);
+  }
+
+  /**
+   * Creates a Prefixed native peer given the base access native handle.
+   * @throws RuntimeException if the base access is unsupported; or if the namespace is not valid
+   */
+  private static native long nativeCreate(String namespace, long accessNativeHandle);
```

Comment:
nit: we usually use (handle, arguments....) order in native methods

Reply:
In `create*` a handle is just an argument, it is not a handle to "this".

Reply:
Okay, I guess it's the matter of taste